### ISSUE TITLE
Request: Stream Compaction / Conditional Reduction

### ISSUE THREAD

**Issue by polarathene** [arrayfire/arrayfire#1818]
Would it be possible to add a method to compact an Array? Similar to `select()` but only returning an Array of true values(rather than replacing false with a scalar or another Arrays values). This appears to be called Steam Compaction.

I raised an [issue on the Rust wrapper repo](https://github.com/arrayfire/arrayfire-rust/issues/141) with a usecase. I link to an nvidia blog article and another article with an improved implementation if you'd like more information and reference(they provide cuda source code on github repo).

I have tried to accomplish this using current AF methods, but they are all quite slow, a CPU equivalent is faster. I am not sure how `where()/locate()` works, when I have a boolean Array of what values I want, I can get the indices by multiplying against a similar array containing values representing their index:

```
[0, 1, 0, 0, 1, 1] *= [0, 1, 2, 3, 4, 5] 
becomes
[0, 1, 0, 0, 4, 5] 
compacted: 
[1, 4, 5]
```

I can then extract the values I want from the data Array. If I don't need indices I can multiply the bool Array against the data Array and compact it to get a reduced Array with the values I need. This transfers to the host much better and can prevent wasting time with AF methods on elements that are no longer of interest.

---

Two versions of a PDF[[1](http://www.cs.princeton.edu/courses/archive/fall13/cos326/lec/23-parallel-scan.pdf)][[2](https://courses.cs.washington.edu/courses/cse332/10sp/lectures/lecture20.pdf)] that briefly discusses this method with an example:

Parallel prefix to the rescue
1.   Use a parallel map to compute a *bit-vector* for true elements
    input  [17, 4, 6, 8, 11, 5, 13, 19, 0, 24]
    bits   [1,  0, 0, 0,  1, 0,  1,  1, 0,  1]
2.   Do parallel-prefix sum on the bit-vector
    bitsum [1,  1, 1, 1,  2, 2,  3,  4, 4,  5]
3.   Use a parallel map to produce the output
    output [17, 11, 13, 19, 24]

```C
output = new array of size bitsum[n-1]
if(bitsum[0]==1) output[0] = input[0];
FORALL(i=1; i < input.length; i++)
    if(bitsum[i] > bitsum[i-1])
        output[bitsum[i]-1] = input[i];
```

---

One more resource if helpful, [a paper on the steam compaction](ftp://ftp.repec.org/opt/ReDIF/RePEc/rau/jisomg/SP12/JISOM-SP12-A20.pdf). See page 6 for another visual and much more detailed explanation, where vowels are extracted from the alphabet via boolean mask.

> The **stream compaction is a common parallel primitive**, an essential building block for many data processing algorithms, whose optimization improves the performance of a wide class of parallel algorithms useful in data processing.

Hopefully this is not too difficult to implement into ArrayFire and could be made available in the near future :) I understand you're all very busy and this may be low value / priority right now. A performant conditional filter / reduction would hopefully benefit many other users :)

**Comment by umar456** [comment]
This can be done easily using indexing, and the where function:

```c++
#include <arrayfire.h>

using namespace af;

int main() {
  array a = randu(10).T();
  array bits = randu(10, b8).T();

  af_print(a);
  af_print(bits);

  // Using indexing
  af_print(a(bits));

  // which is the same as
  af_print(a(where(bits)));

  // Using inequality
  af_print(a(a < .6));
}
```

Results:
```
a
[1 10 1 1]
    0.3990     0.8805     0.7357     0.6055     0.7739     0.2727     0.5652     0.0689     0.8790     0.4668 
bits
[1 10 1 1]
         1          1          1          1          0          0          1          1          1          0 
a(bits)
[1 7 1 1]
    0.3990     0.8805     0.7357     0.6055     0.5652     0.0689     0.8790 
a(where(bits))
[1 7 1 1]
    0.3990     0.8805     0.7357     0.6055     0.5652     0.0689     0.8790 
a(a < .6)
[1 5 1 1]
    0.3990     0.2727     0.5652     0.0689     0.4668 
```

You can read more about indexing in this blog post: https://arrayfire.com/indexing-with-arrayfire/

**Comment by polarathene** [comment]
@umar456 When using `where()/locate()` on elements within the 100mil-1bil range performance is quite slow in comparison to other operations. As stated, transfer time aside, the process is considerably faster on CPU. 

If I don't need indices that `where()/locate()` provides and want to reduce an array without the large delay(over 400x as much as my other operations), from sources I've linked it appears to be much faster with parallelized scan(prefix-sum) filtering on the condition(bool array).

**Comment by polarathene** [comment]
@umar456 When intending to process large datasets for hours, such a delay is really adds up to processing time. Is this feature request going to be disregarded because there are current less efficient approach available? Or can this request still be considered for a future release considering AF is all about parallelism and performance?

**Comment by umar456** [comment]
I am not sure if I understand. We are performing a scan operation to generate these results. Can you show me how you are benchmarking the CPU and GPU operations?

**Comment by polarathene** [comment]
@umar456 I was timing between each method. I've seen that this is not neccessarily accurate due to async calls and JIT? I would have method calls to AF before and after that were fairly consistent in time for each method despite the order. `locate()` and `set_intersect()` methods were ones that would take hundreds of ms to perform, while basic math operations or comparisons were fast. 

Creating a new Array with host data was slow(transfer to GPU) as was transferring back to host, understandable. Creating the Arrays on the GPU seemed fast, but that could have been bad measurement on my part. 

I'm not sure how to utilize a profiler with ArrayFire with nvvp(nvidia visual profiler). After compiling my Rust program and trying to use the profiler with it, it fails to detect anything(a user claimed it worked in past versions on Google Groups for AF).

If the same timing accuracy is incorrect on the Rust wrapper, we do not seem to be provided the equivalent timing method? Would timing around an [eval macro call](http://arrayfire.org/arrayfire-rust/arrayfire/macro.eval.html) work? I believe I've commented out the `locate()` method and all other timings were consistent with how they were before.

**Comment by umar456** [comment]
Here is what I see on my system:

```cpp
#include <arrayfire.h>
#include <string>

using namespace af;


template<typename FN>
void my_timer(FN fun) {
  for(int i = 0; i < 10; i++) {
    auto start = timer::start();
    array out = fun();
    out.eval();
    af::sync();
    printf("%f  ", timer::stop(start));
  }
}

int main() {

  af::setBackend(AF_BACKEND_CPU);
  af::info();

  array a    = randu(1000000);
  array bits = randu(1000000, b8);

  printf("cpu indexing   : ");
  my_timer( [&] {return (array)a(bits);});

  printf("\ncpu where      : ");
  my_timer( [&] {return (array)a(where(bits));});

  printf("\ncpu inequality : ");
  my_timer([&] {return (array)a(a < .5);});

  printf("\n");
  af::setBackend(AF_BACKEND_CUDA);
  af::info();

  array a_cuda    = randu(1000000);
  array bits_cuda = randu(1000000, b8);

  printf("\ncuda indexing   : ");
  my_timer([&] {return (array)a_cuda(bits_cuda);});

  printf("\ncuda where      : ");
  my_timer( [&] {return (array)a_cuda(where(bits_cuda));});

  printf("\ncuda inequality : ");
  my_timer([&] {return (array)a_cuda(a_cuda < .5);});
}
```

Result:

```
ArrayFire v3.5.0 (CPU, 64-bit Mac OSX, build f129a08)
[0] Intel: Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz, 16384 MB, Max threads(8) 
cpu indexing   : 0.030422  0.013068  0.010575  0.011862  0.011186  0.011008  0.011156  0.010809  0.010849  0.014838  
cpu where      : 0.011269  0.010874  0.010714  0.010934  0.012253  0.011868  0.013695  0.013114  0.010596  0.010667  
cpu inequality : 0.020420  0.019574  0.018636  0.023117  0.018471  0.023171  0.018729  0.019264  0.019176  0.018507  
ArrayFire v3.5.0 (CUDA, 64-bit Mac OSX, build b61de89)
Platform: CUDA Toolkit 8, Driver: CUDA Driver Version: 8000
[0] GeForce GT 750M, 2048 MB, CUDA Compute 3.0

cuda indexing   : 0.008161  0.014187  0.008434  0.006817  0.007048  0.006934  0.006747  0.006681  0.006793  0.007013  
cuda where      : 0.006902  0.007980  0.013815  0.007804  0.006787  0.006941  0.006841  0.006688  0.006722  0.006823  
cuda inequality : 0.110227  0.004700  0.004726  0.004665  0.004701  0.004706  0.004733  0.004712  0.004700  0.004691  
```

As you can see the CUDA backend is a good bit faster than the CPU backend. Are you not seeing similar results?

**Comment by polarathene** [comment]
When referring to CPU equivalent of `where()/locate()` I was referring to my host language(Rust) processing an array/vector of the result data from AF, not AF doing the operation on CPU backend.

```Rust
// Assume my conditional kept array vals of power of 2, but 100 million+ multiple bool array from `eq()` to set undesired values as 0
// [0, 0, 256, 0, 128, 0, 0, 0, 128, 512]
// `where()/locate()` could do filter but timed at 400+ ms
// once data is transferred to host, we can filter based on condition, not 0 - about 60ms in comparison
results.into_iter().filter(|&b| b!=0).collect();
// [256, 128, 128, 512] // if indices were needed can easily get with enumerate() on Rust iterator prior to filter
```

I will try create a sample program and time like you have with eval and sync calls(I've not been using these, so perhaps my timing is inaccurate for AF).

**Comment by pavanky** [comment]
@umar456 This is fairly similar to `where` but would replace the last step. Instead of writing out the global indices, it will just write out the original values.

**Comment by polarathene** [comment]
@umar456 Here is the Rust equivalent as best I can translate, @pavanky can probably confirm:

```Rust
#[macro_use(af_print,eval)]
extern crate arrayfire as af;
use af::*;


use std::time::Instant;
macro_rules! my_timer {
    ($label:expr, $($content:tt)*) => (
        for i in 0..10 {
            let start = Instant::now();

            $($content)*

            af::sync(0);
            println!("{} took: {} sec, {} ms", $label, start.elapsed().as_secs(), (start.elapsed().subsec_nanos() as f64 / 1_000_000f64) );
        };
    )
}

fn main() {
    af::info();
    af::set_backend(Backend::DEFAULT); // Having trouble getting CUDA to work despite following instructions for NVVM.


    let dims      = Dim4::new(&[1000, 1000, 1, 1]); //1_000_000
    let a_cuda    = randu::<u32>(dims);
    let bits_cuda = randu::<u8>(dims);


    my_timer!("cuda indexing",
        // Create an AF indexer
        let mut idxrs = Indexer::new();
        // Not sure if batch should be true, might also have to index against col dim(1) too?
        idxrs.set_index(&bits_cuda, 0, Some(true));
        // Apply the array as an index
        let temp = &index_gen(&a_cuda, idxrs);
        // eval macro requires array be assigned to a temporary variable(identifier)
        eval!(
            temp
        );
    );


    my_timer!("cuda locate",
        let mut idxrs = Indexer::new();
        // Can avoid assignment in next update due to bugfix
        let indices = af::locate(&bits_cuda);
        idxrs.set_index(&indices, 0, Some(true));
        let temp = &index_gen(&a_cuda, idxrs);
        eval!(
            temp
        );
    );


    my_timer!("cuda inequality",
        let neq_result = af::lt(&a_cuda, &{u32::max_value()/2}, true);
        let mut idxrs = Indexer::new();
        // Can avoid assignment in next update due to bugfix
        let indices = af::locate(&neq_result);
        idxrs.set_index(&indices, 0, Some(true));
        let temp = &index_gen(&a_cuda, idxrs);
        eval!(
            temp
        );
    );
}
```

Output:
```
ArrayFire v3.4.0 (OpenCL, 64-bit Linux, build 10d9716)
[0] NVIDIA  : GeForce GTX 1070, 8105 MB
cuda indexing took: 0 sec, 1.553349 ms
cuda indexing took: 0 sec, 0.498429 ms
cuda indexing took: 0 sec, 0.487913 ms
cuda indexing took: 0 sec, 0.486778 ms
cuda indexing took: 0 sec, 0.476694 ms
cuda indexing took: 0 sec, 0.473464 ms
cuda indexing took: 0 sec, 0.769965 ms
cuda indexing took: 0 sec, 0.541601 ms
cuda indexing took: 0 sec, 0.487277 ms
cuda indexing took: 0 sec, 0.4888 ms
cuda locate took: 0 sec, 2.888693 ms
cuda locate took: 0 sec, 0.586138 ms
cuda locate took: 0 sec, 0.614405 ms
cuda locate took: 0 sec, 0.611709 ms
cuda locate took: 0 sec, 0.614033 ms
cuda locate took: 0 sec, 0.591388 ms
cuda locate took: 0 sec, 0.599444 ms
cuda locate took: 0 sec, 0.611882 ms
cuda locate took: 0 sec, 0.613893 ms
cuda locate took: 0 sec, 0.595863 ms
cuda inequality took: 0 sec, 1.524836 ms
cuda inequality took: 0 sec, 0.421424 ms
cuda inequality took: 0 sec, 0.430091 ms
cuda inequality took: 0 sec, 0.42137 ms
cuda inequality took: 0 sec, 0.418541 ms
cuda inequality took: 0 sec, 0.422331 ms
cuda inequality took: 0 sec, 0.425306 ms
cuda inequality took: 0 sec, 0.433456 ms
cuda inequality took: 0 sec, 0.409514 ms
cuda inequality took: 0 sec, 0.668451 ms
```

With sync and eval removed:
```
ArrayFire v3.4.0 (OpenCL, 64-bit Linux, build 10d9716)
[0] NVIDIA  : GeForce GTX 1070, 8105 MB
cuda indexing took: 0 sec, 1.095479 ms
cuda indexing took: 0 sec, 0.04915 ms
cuda indexing took: 0 sec, 0.028318 ms
cuda indexing took: 0 sec, 0.038892 ms
cuda indexing took: 0 sec, 0.025994 ms
cuda indexing took: 0 sec, 0.025461 ms
cuda indexing took: 0 sec, 0.038684 ms
cuda indexing took: 0 sec, 0.025391 ms
cuda indexing took: 0 sec, 0.025572 ms
cuda indexing took: 0 sec, 0.040157 ms
cuda locate took: 0 sec, 6.7955 ms
cuda locate took: 0 sec, 0.522479 ms
cuda locate took: 0 sec, 0.55811 ms
cuda locate took: 0 sec, 0.602838 ms
cuda locate took: 0 sec, 0.549322 ms
cuda locate took: 0 sec, 0.560668 ms
cuda locate took: 0 sec, 0.562326 ms
cuda locate took: 0 sec, 0.562115 ms
cuda locate took: 0 sec, 0.559315 ms
cuda locate took: 0 sec, 0.575579 ms
cuda inequality took: 0 sec, 1.580555 ms
cuda inequality took: 0 sec, 0.315871 ms
cuda inequality took: 0 sec, 0.372846 ms
cuda inequality took: 0 sec, 0.376542 ms
cuda inequality took: 0 sec, 0.37198 ms
cuda inequality took: 0 sec, 0.376806 ms
cuda inequality took: 0 sec, 0.375088 ms
cuda inequality took: 0 sec, 0.375404 ms
cuda inequality took: 0 sec, 0.389608 ms
cuda inequality took: 0 sec, 0.375825 ms
```

Mostly the same?

I used this measurement for time on my own code for locate() and the eq() statement before it:

```
test array size: elements: 297_034_400, rows: 59_406_880, cols: 5

my timer macro for eq() (no eval or sync) took: 0 sec, 0.035072 ms
eq() took: 0 sec, 5.534959 ms
eq() took: 0 sec, 5.989262 ms
eq() took: 0 sec, 5.493099 ms
eq() took: 0 sec, 5.467679 ms
eq() took: 0 sec, 6.01276 ms
eq() took: 0 sec, 5.881209 ms
eq() took: 0 sec, 5.513534 ms
eq() took: 0 sec, 5.929341 ms
eq() took: 0 sec, 5.970329 ms
eq() took: 0 sec, 6.051082 ms
af::locate() took: 0 sec, 9.57215 ms
af::locate() took: 0 sec, 3.613201 ms
af::locate() took: 0 sec, 3.654709 ms
af::locate() took: 0 sec, 3.658023 ms
af::locate() took: 0 sec, 3.65451 ms
af::locate() took: 0 sec, 3.654957 ms
af::locate() took: 0 sec, 3.656441 ms
af::locate() took: 0 sec, 3.649532 ms
af::locate() took: 0 sec, 3.655408 ms
af::locate() took: 0 sec, 3.658911 ms
```

The 400ms+ I experienced earlier would have been when I was creating my dataset on the CPU and transferring to AF to put on the GPU, I guess my timer attributed it to locate() for some reason when the 400ms might have been due to the transfer of a couple GB of data that it claimed took 70ms.

Apologies, I assumed locate() was bottlenecking. **It would still be a nice feature to get as @pavanky describes**, at least for Rust which as you can see extracting the indexed results is rather verbose compared to your C++ version.

**Comment by v31l-sys** [comment]
Hi there, is this still being worked on? I'm working with very large arrays (150 million elements each, and create new arrays of that size and in the end will have ended up creating ~117,000 of those arrays).

The issue is that  compacting these arrays is literally magnitudes more expensive to do than something like select() and replacing the values with 0, this is the function I'm using to reduce the arrays by a condition.

```rust
pub fn filter(arr: &Array<i64>, condition: &Array<bool>) -> Array<i64> {
    //let mut temp: Array<bool> = eq(&shiftr(&arr, &44, false), &7, false);
    //af_print!("", condition);
    let x = locate(condition);
    if !x.is_empty() {
        let mut idxr = Indexer::default();
        idxr.set_index(&x, 0, None);
        index_gen(arr, idxr)
    } else {
        Array::<i64>::new_empty(Dim4::new(&[1, 1, 1, 1]))
    }
}
```

It would be really nice to have a select() that only produces an array that satisfies the condition specified by Array<bool>.

**Comment by 9prady9** [comment]
> Apologies, I assumed locate() was bottlenecking. **It would still be a nice feature to get as @pavanky describes**, at least for Rust which as you can see extracting the indexed results is rather verbose compared to your C++ version. 

I have recently added few macros to enable a more terse syntax to index using arrays/sequences - the associated release should be available by end of this month or earlier.

**Comment by 9prady9** [comment]
> Hi there, is this still being worked on? I'm working with very large arrays (150 million elements each, and create new arrays of that size and in the end will have ended up creating ~117,000 of those arrays).
> 
> The issue is that  compacting these arrays is literally magnitudes more expensive to do than something like select() and replacing the values with 0, this is the function I'm using to reduce the arrays by a condition.
> 
> ```rust
> pub fn filter(arr: &Array<i64>, condition: &Array<bool>) -> Array<i64> {
>     let x = locate(condition);
>     if !x.is_empty() {
>         let mut idxr = Indexer::default();
>         idxr.set_index(&x, 0, None);
>         index_gen(arr, idxr)
>     } else {
>         Array::<i64>::new_empty(Dim4::new(&[1, 1, 1, 1]))
>     }
> }
> ```
> 
> It would be really nice to have a select() that only produces an array that satisfies the condition specified by Array<bool>.
> 
> Because of the performance degradation from using this form of filtering / compaction, it ends up being faster to run the code in parallel on a CPU

As the original issue author @polarathene mentioned above, performance wasnt the issue in his case as it turned out. Can you please show us how you are timing your code.

**Comment by v31l-sys** [comment]
To clarify I'm using the OpenCL backend on an RX 580

This is a single CPU thread implementation of what I'm doing just to illustrate:
```rust
let now = Instant::now();
    let mut x_term: Vec<i64> = Vec::<i64>::new();
    unsafe {
        for _i in 0i64..3200000000 { //3.2 billion
            let mut seed = _i + (12 << 44);
            seed = seed.wrapping_mul(0x5DEECE66D).wrapping_add(0xB) & (((1 as i64) << 48) - 1);
            if (seed >> 44) == 5 {
                x_term.push(seed);
            }
        }
    }
    println!("single core time: {}", now.elapsed().as_millis());
```
Which gives me a time of roughly 2.2 seconds on a single thread. If it was implemented in parallel it would obviously scale pretty well, on my system that would mean something like ~275 - 350 milliseconds to completion.

The implementation in ArrayFire is like the following:
```rust
fn calculate_seeds(coordinate: &[Coordinate]) {
    let mut slice_index: i64 = 0;
    let mut proc_count: i16 = 0;
    let mut now = Instant::now();

    //this is basically our for loop
    while slice_index < 0x100000000000 { //n^44

        let bytes_to_allocate: i64 = 150000000; //arrays of i64 150000000 in size

        //this is where the magic is happening
        let mut base: Array<i64> = init_array(bytes_to_allocate, slice_index, 5); //5 just for example
        eval!(&base);

        if slice_index == 0 {
            //set now to the time after ArrayFire initially initializes its array
            now = Instant::now();
        }

        proc_count += 1;
        if proc_count == 21 { //roughly 3.2 billion elements are iterated over 21 loops of 150 million
            println!("proc loop count: {} loop % completion: {} in time: {}", proc_count,
                     ((slice_index + BYTES_TO_ALLOCATE) as f32 / 17592186044416.0) * 100.0, now.elapsed().as_millis());

            mem_info!("arrays processed...");
            break;
        }

        slice_index += 150000000;
    }
```

Which is calling the function:

```rust
pub fn init_array(byte_allocation: i64, current_index: i64, coordinate: i32) -> Array<i64> {
    let dims = Dim4::new(&[byte_allocation as u64, 1, 1, 1]);

    //essentially initialize the array to ((coordinate << 44) + offset) + index (from 0..150000000 in this cast)
    let mut base = add(&constant::<i64>(((coordinate as i64) << 44) + current_index, dims),
                                &range(dims, 0) as &Array<i64>, false);
    base = next(&base, &FORWARD1);
    eval!(&base);

    //base = selectr(&base, &eq(&shiftr(&base, &44, false), &coordinate.rel_z, false), 0.0);
    base = filter(&base, &eq(&shiftr(&base, &44, false), coordinate + 1, false)); //is each (number >> 44) == coordinate + 1? 
    //the math works out to basically only 6.25% of the Array being valid by this condition, meaning it's mostly zeros.

    base = next(&base, &FORWARD1);
    eval!(&base);

    base
}
```

Which calls the function next (||| basically it calculates (number * 0x5DEECE66D + 0xB) & ((1 << 48) - 1) |||).
This does not seem to the issue because without having to do the compaction it can EASILY crunch through nearly 2 trillion values in less than 10 seconds, even if it's not the best implementation.

```rust
pub fn next(arr: &Array<i64>, lcg: &LCG) -> Array<i64>
{
        bitand(
            &add(
                &mul(&*arr, &lcg.multiplier, false),
            &lcg.addend, false),
       &constant::<i64>(lcg.modulus, Dim4::new(&[1, 1, 1, 1])), true)
}
```

The function I'm using for filtering is a couple comments up.

On my system the total time is about ~840 milliseconds, which would make the CPU implementation nearly twice as fast in this case. To be fair my optimization is probably god awful or this is the kind of problem that stream compaction is not really designed for.

Is this the kind of problem where it would be better to offload the initial bulk of the work to the CPU? Is this even a feasible thing to run on a GPU exclusively is the question.

**Comment by 9prady9** [comment]
@v31l-sys Sorry about the delay, I was busy with covid-19 :) still recovering. I have looked at your code and on quick look it seems fine to me, although I would prefer running a standalone version on my local setup to see what chunks of code is taking up more time. Can you please provide me a standalone snippet that I can run. If the code is open source on github already, do point me to that with setup instructions.

Having said that, just reminding(you may already know) that GPUs do have device warm up time, especially with JIT code and functions, the very first call to a function or a set of logic that translates to a JIT kernel on it's first evaluation needs to be compiled to device code at runtime - subsequent calls use the cached device binaries.
